game-core/src/main/java/games/strategy/engine/framework/startup/ui/editors/DiceServerEditor.java

```diff
@@ -2,111 +2,115 @@
+    row++;
+    add(ccLabel, new GridBagConstraints(0, row, 1, 1, 0, 0, GridBagConstraints.WEST, GridBagConstraints.NONE,
+        new Insets(0, 0, bottomSpace, labelSpace), 0, 0));
+    add(ccAddress, new GridBagConstraints(1, row, 2, 1, 1.0, 0, GridBagConstraints.EAST,
+        GridBagConstraints.HORIZONTAL, new Insets(0, 0, bottomSpace, 0), 0, 0));
+    row++;
+    final JLabel gameIdLabel = new JLabel("Game Name:");
+    add(gameIdLabel, new GridBagConstraints(0, row, 1, 1, 0, 0, GridBagConstraints.WEST, GridBagConstraints.NONE,
+        new Insets(0, 0, bottomSpace, labelSpace), 0, 0));
+    add(gameId, new GridBagConstraints(1, row, 2, 1, 1.0, 0, GridBagConstraints.EAST, GridBagConstraints.HORIZONTAL,
+        new Insets(0, 0, bottomSpace, 0), 0, 0));
+    row++;
+    add(testDiceButton, new GridBagConstraints(2, row, 1, 1, 0, 0, GridBagConstraints.EAST, GridBagConstraints.NONE,
         new Insets(0, 0, bottomSpace, 0), 0, 0));
-    setupListeners();
-  }
 
-  /**
-   * Configures the listeners for the gui components.
-   */
-  private void setupListeners() {
-    testDiceyButton.addActionListener(e -> {
-      final PbemDiceRoller random = new PbemDiceRoller(getDiceServer());
+    servers.addItemListener(e -> checkFieldsAndNotify());
+    testDiceButton.addActionListener(e -> {
+      final PbemDiceRoller random = new PbemDiceRoller(createDiceServer());
       random.test();
     });
-    final DocumentListener docListener = new EditorChangedFiringDocumentListener();
-    toAddress.getDocument().addDocumentListener(docListener);
-    ccAddress.getDocument().addDocumentListener(docListener);
+    toAddress.getDocument().addDocumentListener(new TextFieldInputListenerWrapper(this::checkFieldsAndNotify));
+    ccAddress.getDocument().addDocumentListener(new TextFieldInputListenerWrapper(this::checkFieldsAndNotify));
+    gameId.getDocument().addDocumentListener(new TextFieldInputListenerWrapper(this::checkFieldsAndNotify));
   }
 
-  @Override
-  public boolean isBeanValid() {
-    boolean toValid = true;
-    boolean ccValid = true;
-    if (getDiceServer().sendsEmail()) {
-      toValid = validateTextField(toAddress, toLabel, new EmailValidator(false));
-      ccValid = validateTextField(ccAddress, ccLabel, new EmailValidator(true));
-    }
-    final boolean allValid = toValid && ccValid;
-    testDiceyButton.setEnabled(allValid);
+  private void checkFieldsAndNotify() {
+    areFieldsValid();
+    readyCallback.run();
+  }
+
+  public boolean areFieldsValid() {
+    final String toAddressText = toAddress.getText();
+    final boolean toValid = setLabelValid(!toAddressText.isEmpty() && Util.isMailValid(toAddressText), toLabel);
+    final boolean ccValid = setLabelValid(Util.isMailValid(ccAddress.getText()), ccLabel);
+    final boolean serverValid = validateCombobox(servers, serverLabel);
+    final boolean allValid = serverValid && toValid && ccValid;
+    testDiceButton.setEnabled(allValid);
```

Comment:
Possible, yes but not really pretty either because the "side-effect-version" of this method would still look almost the same, just without the boolean return value.